### PR TITLE
feat(settings): add method that saves settings to file

### PR DESCRIPTION
**Pull request by SrGesus**
# Description

Include a summary of the changes here.

## Checklist

- [x] Self-review changes.
- [x] Evaluate impact on the documentation.
- [ ] Ensure test coverage.
- [x] Write new samples.
- [x] Add entry to the changelog's unreleased section.

### CODE REVIEW

**Comment by RiscadoA** [comment]
If we do end up reimplementing the settings resource as a system argument, this would no longer be necessary :thinking: 
You could just have a `Settings::save` method, which ig would be even better because then you don't need to mess with the weird event system we have and it is also synchronous. Opinions?

**Comment by SrGesus** [comment]
I think we wanted events to be able to do something if a setting changes, e.g reload graphics if you change a graphics setting or something and there we will need to use something similar if not events (like an observer idk).
But yes for saving i don't need events, in fact i can do it without events without reimplementing settings

File: engine/src/settings/settings.cpp
Comment (RiscadoA) [inline engine/src/settings/settings.cpp:131]: I think it is probably a better a idea to turn this into a `nlohmann::json`, and then serialize that value with `json.dump()`. Less code duplication.

File: engine/src/settings/settings.cpp
Comment (RiscadoA) [inline engine/src/settings/settings.cpp:135]:
```suggestion
bool Settings::save(std::string_view path)
{
```
I would change this to receive the virtual file system path where to save to, and by default set it to `/settings.json`.
This gives the caller more flexibility. Internally then, you wouldn't mount an archive, you would only keep the last three lines. Since the settings file is already mounted at startup anyway the logic in the beginning is unnecessary

**Review by RiscadoA — CHANGES_REQUESTED** [review]
Thanks for working on this! I would just simplify this a bit before merging.
Also, don't forget the changelog entry :relieved:

**Review by RiscadoA — CHANGES_REQUESTED** [review]
Accidentally dismissed my own review, my bad

File: engine/src/settings/settings.cpp
Comment (SrGesus) [reply engine/src/settings/settings.cpp:131]: I avoided doing that because:
- Serializing the map into a nlohmann::json sounds like an expensive process with a lot of allocations just to then Serialize it again into a string, compared to just iterating the values
- That will add quotation marks to every value :/

**Comment by diogomsmiranda** [comment]
What is the status on this? Is this ready for review?

File: engine/src/settings/settings.cpp
Comment (RiscadoA) [reply engine/src/settings/settings.cpp:131]: Regarding the first point, performance isn't really an issue here. We won't be saving the settings every frame thousands of times! So, IMO readability/maintainability is the priority here.

Regarding the second point, not necessarily; if you do something like `nlohmann::json json = 32`, it dumps you an integer, without the quotation marks. It will only put quotation marks if you do `json = "32"`, so it's just a matter of doing it like you're doing right now - checking if it is a number of a boolean

Comment (SrGesus) [reply engine/src/settings/settings.cpp:131]: But we don't have integers we just have a map of string -> string so every variable is a string

Comment (SrGesus) [reply engine/src/settings/settings.cpp:131]: Ah you're saying making a new json, iterating the map and adding each element individually, converting it into a double or a boolean?

Comment (RiscadoA) [reply engine/src/settings/settings.cpp:131]: Sure, but if it is a number you can just use the return value of `strtod` no? And for the booleans, you can do `json = value == "true";`

**Comment by SrGesus** [comment]
> ready
The settings part should be done now. I just intend on adding a few tests and the changelog.

File: engine/include/cubos/engine/settings/settings.hpp
Comment (RiscadoA) [inline engine/include/cubos/engine/settings/settings.hpp:106]:
```suggestion
        /// @param indent Indent of the Json, -1 for no whitespace.
```

File: engine/include/cubos/engine/settings/settings.hpp
Comment (RiscadoA) [inline engine/include/cubos/engine/settings/settings.hpp:112]:
```suggestion
        /// @return Loaded settings object.
```

File: engine/samples/settings/main.cpp
Comment (RiscadoA) [inline engine/samples/settings/main.cpp:17]:
```suggestion
        settings.save();
```
Maybe omit the arguments here? :thinking: or do you want to showcase them?

File: engine/samples/settings/main.cpp
Comment (RiscadoA) [inline engine/samples/settings/main.cpp:25]: Same as above

File: engine/src/settings/settings.cpp
Comment (RiscadoA) [inline engine/src/settings/settings.cpp:188]: I think you shoud just check if the `stream` returned by `open` is `nullptr`.
E.g., the path may point to a directory instead of a file. In that case, this will succeed, but open will return nullptr

File: engine/src/settings/settings.cpp
Comment (RiscadoA) [inline engine/src/settings/settings.cpp:143]: I think you shoud just check if the `stream` returned by `open` is `nullptr`.
E.g., the path may point to a directory instead of a file. In that case, this will succeed, but open will return nullptr

**Review by RiscadoA — APPROVED** [review]
Other than the comments I pointed out, LGTM!

**Comment by RiscadoA** [comment]
Also don't forget to update the CHANGELOG!

**Review by RodrigoVintem — APPROVED** [review]

**Pull request merged**
